Redis Programmability Documentation (#1725)
redis/redis-doc: topics/programability.md

# Redis Programability

> Programability in Redis allows the user to send functionality (scritps) into Redis that will run on the sever where the data leaves.
> This can reduce network hops and improve overal perfromance.
> It can also enrich the API provided by Redis with a user define API,
> and allow implementing a schema over Redis data structures (updates multiple keys together, keep keys in sync).

Reviewer: ```suggestion
Programability in Redis allows the user to send functionality (scripts) into Redis that will run on the server where the data lives.
This can reduce network hops and improve overall performance.
It can also enrich the API provided by Redis with a user-defined API,
and allow implementing a schema over Redis data structures (update multiple keys together, keep keys in sync).
```

Currently, scripts can only be written in [Lua](https://www.lua.org/).
Redis runs an embeded Lua 5.1 interpreter that is used to run the user scripts.
The full Lua API provided to the user is documented on [Redis Lua API](/topics/lua).

> In addition, Redis provides two ways to run the scripts:

Reviewer: In addition to what? (In addition to simply providing a specification is not additional enough. :D)

```suggestion
Redis provides two ways to run scripts:
```

Redis guarantees that a script is executed in an atomic way:
no other script or Redis command will be executed while a script is being executed.
This semantic is similar to the one of `MULTI` / `EXEC`.
From the point of view of all the other clients the effects of a script are either still not visible or already completed.

However this also means that executing slow scripts is not a good idea.
It is not hard to create fast scripts, as the script overhead is very low,
but if you are going to use slow scripts you should be aware that while the script is running no other client can execute commands.

## Sandbox and maximum execution time

Scripts should never try to access the external system, like the file system or any other system call.
A script should only operate on Redis data and passed arguments.

Scripts are also subject to a maximum execution time (five seconds by default).
This default timeout is huge since a script should usually run in under a millisecond.
The limit is mostly to handle accidental infinite loops created during development.

It is possible to modify the maximum time a script can be executed with millisecond precision,

> either via `redis.conf` or using the CONFIG GET / CONFIG SET command.

Reviewer: ```suggestion
either via `redis.conf` or using the `CONFIG GET` / `CONFIG SET` commands.
```

The configuration parameter affecting max execution time is called `script-time-limit`.

When a script reaches the timeout it is not automatically terminated by Redis since this violates the contract Redis has with the scripting engine to ensure that scripts are atomic.
Interrupting a script means potentially leaving the dataset with half-written data.

> For this reasons when a script executes for more than the specified time the following happens:

Reviewer: ```suggestion
For this reason, when a script executes for more than the specified time the following happens:
```
